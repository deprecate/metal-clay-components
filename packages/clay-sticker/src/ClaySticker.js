import 'clay-icon';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClaySticker.soy.js';

/**
 * Implementation of the Metal Clay Link.
 */
class ClaySticker extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClaySticker.STATE = {
	/**
	 * Render ClayIcon in the ClaySticker element.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?Object|undefined}
	 * @default undefined
	 */
	icon: Config.shapeOf({
		spritemap: Config.string().required(),
		symbol: Config.string().required(),
	}),

	/**
	 * Id to be applied to the sticker element.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Sets the text or HTML to be rendered inside sticker.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?html|string|undefined}
	 * @default undefined
	 */
	label: Config.any(),

	/**
	 * Flag to indicate if the sticker should be positioned on the outside
	 * corners when position is defined.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?bool}
	 * @default false
	 */
	outside: Config.bool().value(false),

	/**
	 * Sticker position if not static. Available positions are `bottom-left`,
	 * `bottom-right`, `top-right`.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?string|undefined}
	 * @default undefined
	 */
	position: Config.oneOf([
		'bottom-left',
		'bottom-right',
		'top-left',
		'top-right',
	]),

	/**
	 * Sticker size. Available sizes are `sm`, `lg`, `xl`.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?string|undefined}
	 * @default undefined
	 */
	size: Config.oneOf(['lg', 'sm', 'xl']),

	/**
	 * Shape of the sticker. Available shapes are `circle`, `rounded`.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?string}
	 * @default rounded
	 */
	shape: Config.oneOf(['circle', 'rounded']).value('rounded'),

	/**
	 * Sticker style. Available sizes are: `danger`, `dark`, `info`, `light`,
	 * `primary`, `secondary`, `success`, `warning`.
	 * @instance
	 * @memberof ClaySticker
	 * @type {?string}
	 * @default primary
	 */
	style: Config.oneOf([
		'danger',
		'dark',
		'info',
		'light',
		'primary',
		'secondary',
		'success',
		'warning',
	]).value('primary'),
};

defineWebComponent('clay-sticker', ClaySticker);

Soy.register(ClaySticker, templates);

export {ClaySticker};
export default ClaySticker;
