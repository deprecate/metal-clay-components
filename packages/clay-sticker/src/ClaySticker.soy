{namespace ClaySticker}

/**
 * This renders a sticker.
 */
{template .render}
	{@param? elementClasses: string}
	{@param? icon: string}
	{@param? id: string}
	{@param? imageAlt: string}
	{@param? imageSrc: string}
	{@param? label: html|string}
	{@param? outside: bool}
	{@param? position: string}
	{@param? shape: string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? style: string}

	{let $attributes kind="attributes"}
		class="sticker
			{sp}sticker-{$style ?: 'primary'}

			{if $position}
				{if $outside}
					{sp}sticker-outside
				{/if}
				{sp}sticker-{$position}
			{/if}

			{if $shape == 'circle'}
				{sp}rounded-circle
			{/if}

			{if $size}
				{sp}sticker-{$size}
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<span {$attributes}>
		{if $icon and $spritemap}
			{call ClayIcon.render}
				{param spritemap: $spritemap /}
				{param symbol: $icon /}
			{/call}
		{elseif $imageSrc}
			<img alt="{$imageAlt}" class="img-fluid" src="{$imageSrc}" />
		{elseif $label}
			{$label}
		{/if}
	</span>
{/template}
