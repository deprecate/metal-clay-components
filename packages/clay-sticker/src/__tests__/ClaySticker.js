import ClaySticker from '../ClaySticker';

let sticker;

describe('ClaySticker', function() {
	afterEach(() => {
		if (sticker) {
			sticker.dispose();
		}
	});

	it('should render the default markup', () => {
		sticker = new ClaySticker();

		expect(sticker).toMatchSnapshot();
	});

	it('should render a sticker with classes', () => {
		sticker = new ClaySticker({
			elementClasses: 'my-custom-class',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a sticker with id', () => {
		sticker = new ClaySticker({
			id: 'myId',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a sticker with icon', () => {
		sticker = new ClaySticker({
			icon: {
				spritemap: 'icons.svg',
				symbol: 'plus',
			},
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a sticker with label', () => {
		sticker = new ClaySticker({
			label: '<span>Doc E. Brown</span>',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a small sticker', () => {
		sticker = new ClaySticker({
			size: 'sm',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a large sticker', () => {
		sticker = new ClaySticker({
			size: 'lg',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a extra large sticker', () => {
		sticker = new ClaySticker({
			size: 'xl',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a circle sticker', () => {
		sticker = new ClaySticker({
			shape: 'circle',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a bottom left positioned sticker', () => {
		sticker = new ClaySticker({
			position: 'bottom-left',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a bottom right positioned sticker', () => {
		sticker = new ClaySticker({
			position: 'bottom-right',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a top left positioned sticker', () => {
		sticker = new ClaySticker({
			position: 'top-left',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a top right positioned sticker', () => {
		sticker = new ClaySticker({
			position: 'top-right',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render an outside bottom left positioned sticker', () => {
		sticker = new ClaySticker({
			outside: true,
			position: 'bottom-left',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render an outside bottom right positioned sticker', () => {
		sticker = new ClaySticker({
			outside: true,
			position: 'bottom-right',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render an outside top left positioned sticker', () => {
		sticker = new ClaySticker({
			outside: true,
			position: 'top-left',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render an outside top right positioned sticker', () => {
		sticker = new ClaySticker({
			outside: true,
			position: 'top-right',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `danger color` sticker', () => {
		sticker = new ClaySticker({
			style: 'danger',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `dark color` sticker', () => {
		sticker = new ClaySticker({
			style: 'dark',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `info color` sticker', () => {
		sticker = new ClaySticker({
			style: 'info',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `light color` sticker', () => {
		sticker = new ClaySticker({
			style: 'light',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `primary color` sticker', () => {
		sticker = new ClaySticker({
			style: 'primary',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `secondary color` sticker', () => {
		sticker = new ClaySticker({
			style: 'secondary',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `success color` sticker', () => {
		sticker = new ClaySticker({
			style: 'success',
		});

		expect(sticker).toMatchSnapshot();
	});

	it('should render a `warning color` sticker', () => {
		sticker = new ClaySticker({
			style: 'warning',
		});

		expect(sticker).toMatchSnapshot();
	});
});
