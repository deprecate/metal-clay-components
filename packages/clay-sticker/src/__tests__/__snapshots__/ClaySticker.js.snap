// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClaySticker should render a \`danger color\` sticker 1`] = `<span class="sticker sticker-danger"></span>`;

exports[`ClaySticker should render a \`dark color\` sticker 1`] = `<span class="sticker sticker-dark"></span>`;

exports[`ClaySticker should render a \`info color\` sticker 1`] = `<span class="sticker sticker-info"></span>`;

exports[`ClaySticker should render a \`light color\` sticker 1`] = `<span class="sticker sticker-light"></span>`;

exports[`ClaySticker should render a \`primary color\` sticker 1`] = `<span class="sticker sticker-primary"></span>`;

exports[`ClaySticker should render a \`secondary color\` sticker 1`] = `<span class="sticker sticker-secondary"></span>`;

exports[`ClaySticker should render a \`success color\` sticker 1`] = `<span class="sticker sticker-success"></span>`;

exports[`ClaySticker should render a \`unstyled\` sticker 1`] = `<span class="sticker sticker-unstyled"></span>`;

exports[`ClaySticker should render a \`warning color\` sticker 1`] = `<span class="sticker sticker-warning"></span>`;

exports[`ClaySticker should render a bottom left positioned sticker 1`] = `<span class="sticker sticker-primary sticker-bottom-left"></span>`;

exports[`ClaySticker should render a bottom right positioned sticker 1`] = `<span class="sticker sticker-primary sticker-bottom-right"></span>`;

exports[`ClaySticker should render a circle sticker 1`] = `<span class="sticker sticker-primary rounded-circle"></span>`;

exports[`ClaySticker should render a extra large sticker 1`] = `<span class="sticker sticker-primary sticker-xl"></span>`;

exports[`ClaySticker should render a large sticker 1`] = `<span class="sticker sticker-primary sticker-lg"></span>`;

exports[`ClaySticker should render a small sticker 1`] = `<span class="sticker sticker-primary sticker-sm"></span>`;

exports[`ClaySticker should render a sticker with classes 1`] = `<span class="sticker sticker-primary my-custom-class"></span>`;

exports[`ClaySticker should render a sticker with icon 1`] = `
<span class="sticker sticker-primary">
  <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus">
    <title>plus</title>
    <use xlink:href="icons.svg#plus"></use>
  </svg>
</span>
`;

exports[`ClaySticker should render a sticker with id 1`] = `<span class="sticker sticker-primary" id="myId"></span>`;

exports[`ClaySticker should render a sticker with image 1`] = `
<span class="sticker sticker-primary">
  <img alt="" class="img-fluid" src="imageurl">
</span>
`;

exports[`ClaySticker should render a sticker with image and imagen alternate text 1`] = `
<span class="sticker sticker-primary">
  <img alt="imageAlt" class="img-fluid" src="imageurl">
</span>
`;

exports[`ClaySticker should render a sticker with label 1`] = `
<span class="sticker sticker-primary">
  <span>Doc E. Brown</span>
</span>
`;

exports[`ClaySticker should render a top left positioned sticker 1`] = `<span class="sticker sticker-primary sticker-top-left"></span>`;

exports[`ClaySticker should render a top right positioned sticker 1`] = `<span class="sticker sticker-primary sticker-top-right"></span>`;

exports[`ClaySticker should render an outside bottom left positioned sticker 1`] = `<span class="sticker sticker-primary sticker-outside sticker-bottom-left"></span>`;

exports[`ClaySticker should render an outside bottom right positioned sticker 1`] = `<span class="sticker sticker-primary sticker-outside sticker-bottom-right"></span>`;

exports[`ClaySticker should render an outside top left positioned sticker 1`] = `<span class="sticker sticker-primary sticker-outside sticker-top-left"></span>`;

exports[`ClaySticker should render an outside top right positioned sticker 1`] = `<span class="sticker sticker-primary sticker-outside sticker-top-right"></span>`;

exports[`ClaySticker should render the default markup 1`] = `<span class="sticker sticker-primary"></span>`;
